### PR TITLE
Support for RM scheduling

### PR DESCRIPTION
**Pull request by hduffddybz**
The main idea of rm scheduling can be described as : Task with shorter period can get higher priority and why using rm scheduling algorithm can shortly see here: http://www.embedded.com/electronics-blogs/beginner-s-corner/4023927/Introduction-to-Rate-Monotonic-Scheduling

an simple example for using rm scheduling:
``` c
#include <rtthread.h>

#define THREAD_STACK_SIZE	512
#define THREAD_TIMESLICE	5
#define PERIOD1				10
#define PERIOD2				20
#define WCET1				3
#define WCET2				3

static rt_rms_t rms1 = RT_NULL, rms2 = RT_NULL, rms3 = RT_NULL, rms4 = RT_NULL, rms5 = RT_NULL, rms6 = RT_NULL;

static void rms1_entry(void *parameter)
{
	while(1)
	{	
		rt_kprintf("rms1_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}
}

static void rms2_entry(void *parameter)
{
	rt_rms_detach(rms1);
	while(1)
	{
		rt_kprintf("rms2_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}

static void rms3_entry(void *parameter)
{
	    rt_kprintf("rm3 aperiodic tasks\n");
}

static void rms4_entry(void *parameter)
{
	    rt_kprintf("rm4 aperiodic tasks\n");
}

static void rms5_entry(void *parameter)
{
	while(1)
	{
		rt_kprintf("rms5_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}

static void rms6_entry(void *parameter)
{
	while(1)
	{
		rt_kprintf("rms6_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}


int rt_application_init()
{
	rms1 = rt_rms_create("rms1", rms1_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD1, WCET1);
	if(rms1 != RT_NULL)
		rt_rms_startup(rms1);
	
	rms2 = rt_rms_create("rms2", rms2_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2, WCET2);
	if(rms2 != RT_NULL)
		rt_rms_startup(rms2);
	
	rms3 = rt_rms_create("rms3", rms3_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, 12, 1);
	if(rms3 != RT_NULL)
		rt_rms_startup(rms3);
	
		rms4 = rt_rms_create("rms4", rms4_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, 12, 1);
	if(rms4 != RT_NULL)
		rt_rms_startup(rms4);
	
	rms5 = rt_rms_create("rms5", rms5_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2 + 1, WCET2);
	if(rms5 != RT_NULL)
		rt_rms_startup(rms5);
	
		rms6 = rt_rms_create("rms6", rms6_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2 + 2, WCET2);
	if(rms6 != RT_NULL)
		rt_rms_startup(rms6);
	
	return 0;
	
}
```

you should use #define RT_USING_RMS in rtconfig.h and rt_rms_idle_init() in startup.c.

So in this way, the fixed priority scheduling algorithm cannot be startup, but please notice that in rms, it also support fixed priority tasks.

but when using rm scheduling, it may bring some problems that I did not fully think it over.

- Since the task create/init/startup interface has been changed. It may hard for some drivers to change its code.

And also, it is the first version of rm scheduling, I did not fully test it(I have to admit that I don't know how to test it more strictly), it has lots of restrictions so it may not possible to merge this code:

- Deletion of rms task resource did not happened in the idle task, so in idle task, it just loop. Why using this just because of the simple implementation.

- When the same period of tasks raise to 3, some taska may not execute ever. It may because of the long execution of clock ISR. So, the same period of rms task should not exceed to 3.

- The period of tasks should limit to 0-255, but if you want to use longer periods, you should set RT_RMS_ACCURACY, for example if you set the RT_RMS_ACCURACY to 10, tasks with period of 10 and 11 has the same priority.Next step, using heap algorithm to replace bit map may be a good choice.

RM调度的主要思想是：越短周期的任务拥有越高级别的优先级，为什么使用RMS调度，可以简单看看这篇文章的介绍：http://www.embedded.com/electronics-blogs/beginner-s-corner/4023927/Introduction-to-Rate-Monotonic-Scheduling

一个简单使用RMS的例子：
``` c
#include <rtthread.h>

#define THREAD_STACK_SIZE	512
#define THREAD_TIMESLICE	5
#define PERIOD1				10
#define PERIOD2				20
#define WCET1				3
#define WCET2				3

static rt_rms_t rms1 = RT_NULL, rms2 = RT_NULL, rms3 = RT_NULL, rms4 = RT_NULL, rms5 = RT_NULL, rms6 = RT_NULL;

static void rms1_entry(void *parameter)
{
	while(1)
	{	
		rt_kprintf("rms1_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}
}

static void rms2_entry(void *parameter)
{
	rt_rms_detach(rms1);
	while(1)
	{
		rt_kprintf("rms2_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}

static void rms3_entry(void *parameter)
{
	    rt_kprintf("rm3 aperiodic tasks\n");
}

static void rms4_entry(void *parameter)
{
	    rt_kprintf("rm4 aperiodic tasks\n");
}

static void rms5_entry(void *parameter)
{
	while(1)
	{
		rt_kprintf("rms5_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}

static void rms6_entry(void *parameter)
{
	while(1)
	{
		rt_kprintf("rms6_entry works!, %d\n", rt_tick_get());
		rt_rms_end_cycle();
	}	
}


int rt_application_init()
{
	rms1 = rt_rms_create("rms1", rms1_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD1, WCET1);
	if(rms1 != RT_NULL)
		rt_rms_startup(rms1);
	
	rms2 = rt_rms_create("rms2", rms2_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2, WCET2);
	if(rms2 != RT_NULL)
		rt_rms_startup(rms2);
	
	rms3 = rt_rms_create("rms3", rms3_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, 12, 1);
	if(rms3 != RT_NULL)
		rt_rms_startup(rms3);
	
		rms4 = rt_rms_create("rms4", rms4_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, 12, 1);
	if(rms4 != RT_NULL)
		rt_rms_startup(rms4);
	
	rms5 = rt_rms_create("rms5", rms5_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2 + 1, WCET2);
	if(rms5 != RT_NULL)
		rt_rms_startup(rms5);
	
		rms6 = rt_rms_create("rms6", rms6_entry, RT_NULL,
						THREAD_STACK_SIZE, THREAD_TIMESLICE, PERIOD2 + 2, WCET2);
	if(rms6 != RT_NULL)
		rt_rms_startup(rms6);
	
	return 0;
	
}
```
要想使用RMS，必须在你的bsp包的rtconfig.h中定义#define RT_USING_RMS，并在startup.c中定义rt_rms_idle_init()

注意到如果使用了RMS，原先的FP（fixed priority）调度就不能被启动了，但是在RMS调度器里面同样支持固定优先级的任务。

使用RMS可能会带来一些问题，我并没有详细的思考过：

- 由于任务的创建，初始化，启动等接口改变了，对于一些已有代码（包括一些驱动，组件等等）的改变会不会太大？

另外这是第一个版本的RM，可能并不稳定，我也没有做过充分的测试（我也不清楚该如何充分的测试？maybe 示波器观察任务的切换时间，任务的激活时间？）代码仍有一些限制所以并不一定合适合并到主干

- 固定优先级的任务删除直接在rt_rms_exit中完成，rm任务删除是在一个周期结束时（这么做的原因可以看issue 239），所以idle task只是在空转

- 当相同周期的任务增加到3时，这几个任务的其中一个会无法被调度，产生的原因可能是因为定时超时函数处理时间过长，所以不要使用超过3个相同周期的任务

- 任务的周期理应限制在0-255，当然要使用更长的周期，你必须定义RT_RMS_ACCURACY，当RT_RMS_ACCURACY为10时，周期为10和11的任务优先级是一样的，这是由于ready queue使用的是bit map的算法所限制的，下一步使用堆来组织ready queue也许是个不错的思路。

欢迎提供意见 继续修改

### CODE REVIEW

**Comment by BernardXiong** [comment]
Good work for the first step. 
* Firstly, please use the same interface as current scheduler.
* You can extend some new RMS API.

**Comment by hduffddybz** [comment]
I can offer some data which I test recently.
According to the Liu & Layland classical paper(1973),the schedulability test for rms is:

rms
the ci stands for worst-case execution time and ti stands for period of tasks.
the data I offered is organized as (ci, ti), In addition, I let the rms task run as long as the wcet.

-  (4, 25),(4,15),(4, 18),(4,20), task(4,20) miss its deadline and delete it,all things will do well according to the paper, so do the reality is.

- (2, 6), (2, 7), (2, 8), (2, 20), task(2,8) and (2,20) miss its deadline and delete them, all things do well .......

- (1, 20), (1, 21), (1, 22), (1, 23), (1, 24), (1, 25),(1, 26),(1, 27), (1, 28), (1, 29) all things do well according to the paper, so do the reality is.

Next time I will generate some random tasks and test it whether it can be schedulable.

**Comment by hduffddybz** [comment]
跟熊大沟通了下，目前的问题大概如下：接口改变很大，在任务初始化，创建，删除等等接口中与FP并不相同（在rm中必须添加period和wcet等字段），所以上层应用程序根本无法使用（该问题存在的原因在于只启动了一个调度器）。所以可能的解决办法在于让rm和fp调度器都启动，rm用rm的接口，fp用fp的接口。当然简单的采取rm调度完再调度fp或者fp调度完了之后再执行rm并不合理，所以可能的解决方法是在原来调度器基础上增加一个抽象层，该抽象层负责调度器的分发，决定当前时刻需要运行什么调度器。maybe a hardwork!

File: src/rms.c [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+        rms->thread->stat = RT_RMS_IDLE;
+
+        /* insert into idle queue */        
+        rt_list_insert_before(&(rt_rms_idle_table[rms->thread->current_priority]),
+                            &(rms->rlist));
+    
+        rt_hw_interrupt_enable(level);  
+    }
+    else
+    {
```
Comment (grissiom) [inline src/rms.c]: The code will take this branch when deadline == tick, am I right?

**Comment by hduffddybz** [comment]
@grissiom When the rms task end one period's work,it should call rt_end_cycle() manually, if current tick shorter than the deadline, it should put it into  idle queue, and this task will be activated in the next period comes.If current tick equal to the deadline,just change the deadline and status of this task.

**Pull request closed without merging**